### PR TITLE
Support externally managed secret keys

### PR DESCRIPTION
**Pull request by kasperisager**
This pull request introduces support for externally managed secret keys to allow a relaying DHT node to act on behalf of another relayed node. A summary of the changes:

- `DHT.createServer()` and `DHT.connect()` now accept two new options, `createHandshake` and `createSecretStream`, to allow callers to perform Noise handshaking and stream setup themselves. To ensure that the relaying DHT node can still hole punch on behalf of the relayed node, `handshake.final()` is now required to return a `holepunchSecret` key, but may leave out the Noise keys.

- `server.listen()` now accepts two new options, `signAnnounce` and `signUnannounce`, to allow callers to perform signing themselves.

- `DHT.lookupAndUnannounce()`, `DHT.unannounce()`, and `DHT.announce()` now accept a new option, `signUnannounce`, to allow callers to perform signing themselves. `DHT.announce()` also accepts a `signAnnounce` option and `DHT.mutablePut()` accepts a `signMutable` option.

### CODE REVIEW

File: lib/noise-wrap.js [hunk 1]
```diff
@@ -37,11 +38,16 @@ module.exports = class NoiseWrap {
...
+    const holepunchCapability = Buffer.allocUnsafe(32)
+
+    sodium.crypto_generichash(holepunchCapability, this.handshake.hash, NS.PEER_HOLEPUNCH)
+
     return {
       isInitiator: this.isInitiator,
       publicKey: this.keyPair.publicKey,
       remotePublicKey: this.remotePublicKey,
       remoteId: NoiseSecretStream.id(this.handshake.hash, !this.isInitiator),
+      holepunchCapability,
```
Comment (mafintosh) [inline lib/noise-wrap.js:50]: holepunchSecret

**Comment by kasperisager** [comment]
@davidmarkclements The tests are failing on Windows as of Brittle 2.0.2. Do you know what might be causing this?

**Comment by kasperisager** [comment]
@mafintosh This is ready for another look.

File: lib/server.js [hunk 2]
```diff
@@ -105,38 +114,48 @@ module.exports = class Server extends EventEmitter {
...
-    const fw = this.firewall(handshake.remotePublicKey, remotePayload, clientAddress)
-
     const hs = {
       round: 0,
       reply: null,
       pair: null,
-      protocols: remotePayload.protocols & this._protocols,
-      firewalled: (!fw || !fw.then) ? Promise.resolve(fw) : fw.catch(toTrue),
+      protocols: this._protocols,
+      firewalled: false,
```
Comment (mafintosh) [inline lib/server.js:126]: init this to true, just to be on the safe side

File: lib/noise-wrap.js [hunk 3]
```diff
@@ -16,32 +17,30 @@ module.exports = class NoiseWrap {
...
-    let payload = null
-
-    try {
-      payload = c.decode(messages.noisePayload, this.handshake.recv(buf))
-    } catch {
-      return null
-    }
-
+  async recv (buf) {
+    const payload = c.decode(messages.noisePayload, this.handshake.recv(buf))
```
Comment (mafintosh) [inline lib/noise-wrap.js:26]: nice, good change!

File: lib/announcer.js [hunk 4]
```diff
@@ -6,18 +6,20 @@ const Persistent = require('./persistent')
...
 module.exports = class Announcer {
-  constructor (dht, keyPair, target) {
+  constructor (dht, keyPair, target, record, opts = {}) {
     this.dht = dht
     this.keyPair = keyPair
     this.target = target
     this.relays = []
     this.stopped = false
-    this.record = c.encode(m.peer, { publicKey: this.keyPair.publicKey, relayAddresses: [] })
+    this.record = record
```
Comment (mafintosh) [inline lib/announcer.js:15]: is this change still needed now? otherwise let's move it to another pr or just keep it as it was

File: index.js [hunk 5]
```diff
@@ -208,11 +211,13 @@ class HyperDHT extends DHT {
...
 
   async mutablePut (keyPair, value, opts = {}) {
+    const signMutable = opts.signMutable || Persistent.signMutable
+
     const target = Buffer.allocUnsafe(32)
     sodium.crypto_generichash(target, keyPair.publicKey)
 
     const seq = opts.seq || 0
-    const signature = Persistent.signMutable(seq, value, keyPair.secretKey)
+    const signature = signMutable(seq, value, keyPair.secretKey)
```
Comment (mafintosh) [inline index.js:220]: this needs async support doesnt it?

File: index.js [hunk 6]
```diff
@@ -145,7 +148,7 @@ class HyperDHT extends DHT {
         signature: null
       }
 
-      ann.signature = Persistent.signAnnounce(target, token, from.id, ann, keyPair.secretKey)
+      ann.signature = signAnnounce(target, token, from.id, ann, keyPair.secretKey)
```
Comment (mafintosh) [inline index.js:151]: async now? note the commit method can be async to help with that (it expects a promise returned)

File: index.js [hunk 7]
```diff
@@ -114,7 +115,7 @@ class HyperDHT extends DHT {
         signature: null
       }
 
-      unann.signature = Persistent.signUnannounce(target, token, from.id, unann, keyPair.secretKey)
+      unann.signature = signUnannounce(target, token, from.id, unann, keyPair.secretKey)
```
Comment (mafintosh) [inline index.js:118]: async also yea? note that this is doable if you just move this logic to a _requestUnnnnounce method or something that is async and call that instead of dht.request below (the unannounces is a list of promises)

File: lib/connect.js [hunk 8]
```diff
@@ -17,7 +17,7 @@ module.exports = function connect (dht, publicKey, opts = {}) {
     dht,
     round: 0,
     target: hash(publicKey),
-    handshake: new NoiseWrap(keyPair, publicKey),
+    handshake: (opts.handshake || defaultHandshake)(keyPair, publicKey),
```
Comment (mafintosh) [inline lib/connect.js:20]: nit: createHandshake?

File: lib/connect.js [hunk 9]
```diff
@@ -182,7 +182,7 @@ async function connectThroughNode (c, node) {
   const { serverAddress, clientAddress, relayed, noise } = await c.dht._router.peerHandshake(c.target, { noise: c.request }, node.from)
   if (isDone(c) || c.connect) return
 
-  const payload = c.handshake.recv(noise)
+  const payload = await c.handshake.recv(noise)
   if (!payload) return
```
Comment (mafintosh) [inline lib/connect.js:186]: check isDone(c) here also like above, since this follows an await now

File: lib/connect.js [hunk 10]
```diff
@@ -170,7 +170,7 @@ async function connectThroughNode (c, node) {
     const addr = c.dht._sockets.remoteServerAddress()
 
     c.firewall = c.dht.firewalled ? FIREWALL.UNKNOWN : FIREWALL.OPEN
-    c.request = c.handshake.send({
+    c.request = await c.handshake.send({
```
Comment (mafintosh) [inline lib/connect.js:173]: this needs an extra flag on c now, requesting: true|false, because we guard above using the c.request property if we should issue the handshake at all, and the await breaks that

File: lib/noise-wrap.js [hunk 11]
```diff
@@ -22,26 +23,24 @@ module.exports = class NoiseWrap {
...
     return payload
   }
 
   final () {
     if (!this.handshake.complete) throw new Error('Handshake did not finish')
 
+    const holepunchSecret = Buffer.allocUnsafe(32)
+
+    sodium.crypto_generichash(holepunchSecret, this.handshake.hash, NS.PEER_HOLEPUNCH)
+
```
Comment (mafintosh) [inline lib/noise-wrap.js:37]: just a mental note that we might want to have a static function or someway of generating this outside the module now we need it returned from final :)

File: lib/server.js [hunk 12]
```diff
@@ -91,6 +94,8 @@ module.exports = class Server extends EventEmitter {
       onpeerholepunch: this._onpeerholepunch.bind(this)
     })
 
+    this._listening = true
```
Comment (mafintosh) [inline lib/server.js:97]: you need to set this to false also behlow in the catch

File: lib/server.js [hunk 13]
```diff
@@ -105,38 +110,48 @@ module.exports = class Server extends EventEmitter {
...
     }
-    if (this.closed) {
+
+    hs.protocols &= remotePayload.protocols
+
+    try {
+      hs.firewalled = await this.firewall(handshake.remotePublicKey, remotePayload, clientAddress)
+    } catch (err) {
+      safetyCatch(err)
+      hs.firewalled = true
```
Comment (mafintosh) [inline lib/server.js:145]: not needed now that it's initialised to true

**Comment by kasperisager** [comment]
I broke the tests 💥

File: lib/connect.js [hunk 14]
```diff
@@ -165,12 +166,13 @@ async function findAndConnect (c, opts) {
...
 async function connectThroughNode (c, node) {
-  if (!c.request) {
+  if (!c.requesting) {
     // If we have a stable server address, send it over now
     const addr = c.dht._sockets.remoteServerAddress()
 
     c.firewall = c.dht.firewalled ? FIREWALL.UNKNOWN : FIREWALL.OPEN
-    c.request = c.handshake.send({
+    c.requesting = true
+    c.request = await c.handshake.send({
```
Comment (mafintosh) [inline lib/connect.js:175]: Last thing, we should check isDone after the send also
